File: Project1/Python/SelectionSort.py
"""
 Name: Alex Bryant
 ID: 2
 Professor: Dr. Dai
 Course: CS416 - Software Engineering
 Date: Feb 19, 2026

 File: SelectionSort.py

 """

def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i+1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

print(selection_sort([64, 34, 25, 12, 22, 11, 90]))

File: Project1/Go/SelectionSort.go
/*
 * Name: Abdelrahman Ahmed
 * Professor: Dr. Dai
 * Course: CS416 - Software Engineering
 * Date: Feb 17, 2026
 *
 * File: SelectionSort.go
 * Description:
 *   Implements Selection Sort by repeatedly selecting the smallest remaining element and placing it into its final position.
 *   It has O(n^2) time complexity and performs well only for small arrays due to many comparisons.
 */

package main

import "fmt"

// SelectionSort sorts the slice in ascending order (in-place).
func SelectionSort(a []int) {
	n := len(a)
	for i := 0; i < n-1; i++ {
		minIdx := i
		for j := i + 1; j < n; j++ {
			if a[j] < a[minIdx] {
				minIdx = j
			}
		}
		a[i], a[minIdx] = a[minIdx], a[i]
	}
}
func main() {
	test := []int{9, 3, 5, 1, 8, 2, 7, 4, 6}
	fmt.Println("Before:", test)
	SelectionSort(test)
	fmt.Println("After: ", test)
}
